ludwig/hyperopt/utils.py

```diff
@@ -51,3 +54,90 @@ def should_tune_preprocessing(config):
         if f"{PREPROCESSING}." in param_name:
             return True
     return False
+
+
+def parameter_to_dict(name, value):
+    if name == ".":
+        # Parameter name ".", means top-level config
+        return value
+
+    parameter_dict = {}
+    curr_dict = parameter_dict
+    name_list = name.split(".")
+    for i, name_elem in enumerate(name_list):
+        if i == len(name_list) - 1:
+            curr_dict[name_elem] = value
+        else:
+            name_dict = curr_dict.get(name_elem, {})
+            curr_dict[name_elem] = name_dict
+            curr_dict = name_dict
+    return parameter_dict
+
+
+def feature_list_to_dict(config: Dict[str, Any]) -> Dict[str, Any]:
+    input_features_dict = {}
+    for feature in config[INPUT_FEATURES]:
+        input_features_dict[feature[NAME]] = feature
+
+    output_features_dict = {}
+    for feature in config[OUTPUT_FEATURES]:
+        output_features_dict[feature[NAME]] = feature
+
+    config = copy.copy(config)
+    config[INPUT_FEATURES] = input_features_dict
+    config[OUTPUT_FEATURES] = output_features_dict
+    return config
+
+
+def feature_dict_to_list(config: Dict[str, Any]) -> Dict[str, Any]:
+    # This works because Python dicts are order-preserving, so we do not need to
+    # do anything special to map from a key in the dict to an index in a list
+    input_features_list = []
+    for feature in config[INPUT_FEATURES].values():
+        input_features_list.append(feature)
+
+    output_features_list = []
+    for feature in config[OUTPUT_FEATURES].values():
+        output_features_list.append(feature)
+
+    config = copy.copy(config)
+    config[INPUT_FEATURES] = input_features_list
+    config[OUTPUT_FEATURES] = output_features_list
+    return config
+
+
+def substitute_parameters(
+    config: Dict[str, Any],
+    parameters: Dict[str, Any],
+):
+    """Update Ludwig config with parameters sampled from the Hyperopt sampler."""
+
+    # Collect the sets of names for each feature grouping so we can map feature names to
+    # groups
+    input_feature_names = {feature[NAME] for feature in config[INPUT_FEATURES]}
+    output_feature_names = {feature[NAME] for feature in config[OUTPUT_FEATURES]}
+
+    # Features in the user config are provided as a list, but in hyperopt we reference
+    # features by name, so convert temporarily to a dict to simplify the mergep process.
+    config = feature_list_to_dict(config)
+
+    # Merge parameters into the user configuration in order. As such, if there are conflicting
+    # params, the later params will take precedence.
+    for name, value in parameters.items():
```

Comment:
One thought about updating the config object with sampled values incrementally while using the custom config parameter "." is that the order of parameters makes a difference, which may matter depending on how the user sets things up. 

For e.g., if there's a `trainer.learning_rate` within "." that also has its own top level parameter like "trainer.learning_rate" outside of ".",  then essentially where the "trainer.learning_rate" falls (before or after the ".") will affect the final value that's picked for that particular trial. 

So if the Ludwig user sets up their parameter space like you have in your tests, then there's no problem, but if they do it slightly differently like in the example above, then it's something that we should make sure they're aware of. Ideally, the guideline should be that they always add "." first and then add config-specific parameters afterwards. 

Just something that might be worth calling out in the docs for 0.6 once this is merged in!

Reply:
Created a task for documenting hyperopting over nested configs: https://github.com/ludwig-ai/ludwig-docs/issues/207, I can take a stab at adding documentation for it and can add you as a reviewer. Should we also call this feature out in 0.6?

Reply:
We may want to hold off on this feature for v0.6, as it's not a bug fix and the branch has already been cut. I think it's fine to wait for v0.7 here, as we don't have plans to use it immediately, and can give it some time to cook before officially releasing it.

I's definitely important that we call it in the docs to parameter ordering consideration you raised. There are some cool things you can do with this, but the user needs to know what they're doing.

Reply:
@tgaddair Got it! In that case, I will need to re-open [this issue](https://github.com/ludwig-ai/ludwig/issues/2243) and get it merged into the 0.6 release branch before Thursday since we want this capability for default parameters to work correctly in 0.6 and there's currently insufficient support. 

Alternatively, we can merge this for now but only call out the defaults feature without mentioning being able to hyperopt over subsection configs.

Let me know what you think is the best way.

Reply:
Ah, interesting. Yeah in that case I'm fine merging this into the 0.6 branch. We can also update the docs to include the nested configs stuff, since it's there.